`rustc_mir_transform` cleanups, round 2

nnethercote opened the pull request:
More cleanups in the style of #129738.

r? @cjgillot

nnethercote committed:
Reduce visibilities, and add `warn(unreachable_pub)`.

Lots of unnecessary `pub`s in this crate. Most are downgraded to
`pub(super)`, though some don't need any visibility.

nnethercote committed:
Factor out some more repetitive code.

nnethercote committed:
Streamline `AbortUnwindingCalls`.

Currently it constructs two vectors `calls_to_terminated` and
`cleanups_to_remove` in the main loop, and then processes them after the
main loop. But the processing can be done in the main loop, avoiding the
need for the vectors.

nnethercote committed:
Simplify `verify_candidate_branch`.

Let chains are perfect for this kind of function.

nnethercote committed:
Remove an unnecessary `continue`.

Nothing comes after it within the loop.

nnethercote committed:
Remove some unnecessary dereferences.

nnethercote committed:
Make `CallSite` non-`Copy`.

It doesn't need to be, and it's 72 bytes on 64-bit platforms, which is
fairly large.

nnethercote committed:
Use `LocalDecls` in a couple of places.

It's nicer than the `IndexVec` type.

nnethercote committed:
Improve consistency in `LowerIntrinsics`.

In some cases `target` and `arg` are obtained fallibly, and in some
cases they are obtained infallibly. This commit changes them all to
infallible.

cjgillot commented on compiler/rustc_mir_transform/src/lower_intrinsics.rs:
```diff
@@ -35,20 +35,19 @@ impl<'tcx> MirPass<'tcx> for LowerIntrinsics {
-                            });
-                            terminator.kind = TerminatorKind::Goto { target };
-                        }
+                        let target = target.unwrap();
```
I'm a bit split on this commit. This could backfire in case we have a pathological input where MIR builder concluded that we don't need a target block, although we need one.

nnethercote replied:
The existing code uses a mix of fallible and infallible. Can you see a reason for this split? To me it wasn't clear, it looked arbitrary, so I tried making it all infallible and the test suite passed...

cjgillot replied:
The only reason I see is personal taste of the person writing the code.
`target` can only be `None` if the returned type is uninhabited. If the type is known, which happens a lot for intrinsics, we can unwrap without any issue. If the type depends on the user code, we can't.

nnethercote replied:
All the intrinsics have a non-`!` return type: `forget`, `{add,sub,mul}_with_overflow`, `size_of`, `min_align_of`, and `discriminant_value`. So I think this change is safe.

nnethercote commented on compiler/rustc_mir_transform/src/abort_unwinding_calls.rs:
```diff
@@ -50,9 +50,7 @@ impl<'tcx> MirPass<'tcx> for AbortUnwindingCalls {
-        let mut calls_to_terminate = Vec::new();
-        let mut cleanups_to_remove = Vec::new();
-        for (id, block) in body.basic_blocks.iter_enumerated() {
+        for block in body.basic_blocks.as_mut() {
```
Good catch. I will change it to `as_mut_preserves_cfg`.

nnethercote replied:
I have switched to `as_mut_preserves_cfg`, but I'm not certain if it's correct.

When you say "changes behaviour", clearing the caches might hurt perf, but won't have any visibile functional effect, right? But switching to `as_mut_preserves_cfg` could have a visible effect. I can just remove this commit if necessary.

nnethercote replied:
I could just go back to `as_mut()` and let the cache be cleared, and measure the perf effect. I suspect it will be negligible.

cjgillot replied:
Using this is actually unsound, as you are modifying the CFG.
You need to call `body.basic_blocks.invalidate_cfg_caches()` explicitly if you use this.
3 options:
- keep as-is;
- use `as_mut()` and potentially clear the cache -> probably not an issue, as this pass runs just after `ElaborateDrops` which changes the CFG a lot, so we may have nothing in cache;
- use `as_mut_preserves_cfg()` and call `invalidate_cfg_caches()` explicitly -> I'm not super fond, as that's a footgun.

nnethercote replied:
I measured and found that the cache is always totally empty when this code path is hit. So I will convert back to `as_mut` and I think that should be good enough.

nnethercote commented on compiler/rustc_mir_transform/src/early_otherwise_branch.rs:
```diff
@@ -310,42 +310,28 @@ fn verify_candidate_branch<'tcx>(
+        && let (discr_place, Rvalue::Discriminant(from_place)) = &**boxed
+        && *from_place == place
+        // ...make that assignment to a local
+        && discr_place.projection.len() == 0
```
Done.

nnethercote commented:
I have updated: I switched to `as_mut_preserves_cfg`, but I'm not sure about it's correctness. I replaced the `len == 0` comparison. I haven't change the `target` unwrapping. Let me know what you think. Thanks!

bors commented:
:umbrella: The latest upstream changes (presumably #129936) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #130016) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote committed:
Use `let`/`else` to de-indent `ElaborateBoxDerefs::run_pass`.

nnethercote commented:
I rebased. Now that I've gone back to `as_mut`, I think this should be ready to merge.

bors commented:
:pushpin: Commit 54459536591630d33a864e89520638c52d8bcf56 has been approved by `cjgillot`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=cjgillot 54459536591630d33a864e89520638c52d8bcf56 -->
<!-- homu: {"type":"Approved","sha":"54459536591630d33a864e89520638c52d8bcf56","approver":"cjgillot","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
